xblock/scorable.py

```diff
@@ -0,0 +1,131 @@
+
+        ScorableXBlockMixin.calculate_score(self, score)
+            Calculate a new score based on the state of the block.  Must not
+            update any values on the block.
+    """
+
+    def rescore(self, only_if_higher):
+        """
+        Calculate a new raw score and save it to the block.  If only_if_higher
+        is True and the score didn't improve, keep the existing score.
+
+        Raises a TypeError if the block cannot be scored.
+        Raises a ValueError if the user has not yet completed the problem.
+        """
+
+        _ = self.runtime.service(self, 'i18n').ugettext
+        resultlog_prefix = _('Rescore result:')
+
+        if not self.allows_rescore():
+            error_msg = _('Problem does not support rescoring: {}').format(self.location)
+            raise TypeError(error_msg)
+
+        original_score = self.get_score()
+        if original_score is None:
+            error_msg = _('Cannot rescore unanswered problem: {}').format(self.location)
+            raise ValueError(error_msg)
+
+        new_score = self.calculate_score()
+
+        self.set_score(new_score)
+        self._publish_grade(only_if_higher)
+        if only_if_higher:
+            only_if_higher_msg = _('Rescore only if score improved.')
+        else:
+            only_if_higher_msg = _('Rescore unconditionally.')
+        result_msg = _('Original score: {}. New score: {}. {}').format(
+            original_score,
+            new_score,
+            only_if_higher_msg
+        )
+        log.info('{} {}'.format(resultlog_prefix, result_msg))
+
+    def allows_rescore(self):
+        """
+        Boolean value: Can this problem be rescored?
+
+        Subtypes may wish to override this if they need conditional support for
+        rescoring.
+        """
+        return True
+
+    def get_score(self):
+        """
+        Return a raw score already persisted on the XBlock.
+
+        Returns:
+            Score(raw_earned=float, raw_possible=float)_
+        """
+        raise NotImplementedError
+
+    def set_score(self, score):
+        """
+
+        Persist a score to the XBlock.
+
+        The score is a named tuple with a raw_earned attribute and a
+        max_possible attribute, reflecting the raw earned score and the maximum
+        raw score the student could have earned respectively.
+
+        Arguments:
+            score: Score(raw_earned=float, raw_possible=float)
+
+        Returns:
+            None
+        """
+        raise NotImplementedError
+
+    def calculate_score(self):
+        """
+        Calculate a new raw score based on the state of the problem.
+        This method should should not modify the state of the
+        XBlock.
+
+        Returns:
+            Score(raw_earned=float, raw_possible=float)
+        """
+        raise NotImplementedError
+
+    def _publish_grade(self, only_if_higher=None):
+        """
+        Publish a grade to the runtime.
+        """
+        score = self.get_score()
```

Comment:
Do not expect `get_score` to be a fast operation.  As it is right now, it is getting called 3-times (!).  Once here, once in the caller of this method, once again in the grades layer.